src/hb-ot-shape-complex-arabic-pua.hh

```diff
@@ -0,0 +1,385 @@
+  {0xF251u, 0xF252u, 0xF253u, 0xF254u}, /* U+062B ARABIC LETTER THEH */
+  {0xF255u, 0xF256u, 0xF257u, 0xF258u}, /* U+062C ARABIC LETTER JEEM */
+  {0xF259u, 0xF25Au, 0xF25Cu, 0xF260u}, /* U+062D ARABIC LETTER HAH */
+  {0xF261u, 0xF262u, 0xF263u, 0xF264u}, /* U+062E ARABIC LETTER KHAH */
+  {0x0000u, 0x0000u, 0xF266u, 0xF265u}, /* U+062F ARABIC LETTER DAL */
+  {0x0000u, 0x0000u, 0xF268u, 0xF267u}, /* U+0630 ARABIC LETTER THAL */
+  {0x0000u, 0x0000u, 0xF26Au, 0xF269u}, /* U+0631 ARABIC LETTER REH */
+  {0x0000u, 0x0000u, 0xF26Cu, 0xF26Bu}, /* U+0632 ARABIC LETTER ZAIN */
+  {0xF26Du, 0xF26Eu, 0xF26Fu, 0xF270u}, /* U+0633 ARABIC LETTER SEEN */
+  {0xF271u, 0xF272u, 0xF273u, 0xF274u}, /* U+0634 ARABIC LETTER SHEEN */
+  {0xF275u, 0xF276u, 0xF277u, 0xF278u}, /* U+0635 ARABIC LETTER SAD */
+  {0xF279u, 0xF27Au, 0xF27Cu, 0xF27Eu}, /* U+0636 ARABIC LETTER DAD */
+  {0xF27Fu, 0xF2F1u, 0xF2A1u, 0xF2A2u}, /* U+0637 ARABIC LETTER TAH */
+  {0xF2A3u, 0xF2A4u, 0xF2A5u, 0xF2A6u}, /* U+0638 ARABIC LETTER ZAH */
+  {0xF2A7u, 0xF2A8u, 0xF2A9u, 0xF2AAu}, /* U+0639 ARABIC LETTER AIN */
+  {0xF2ABu, 0xF2ACu, 0xF2ADu, 0xF2AEu}, /* U+063A ARABIC LETTER GHAIN */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063B ARABIC LETTER KEHEH WITH TWO DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063C ARABIC LETTER KEHEH WITH THREE DOTS BELOW */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063D ARABIC LETTER FARSI YEH WITH INVERTED V */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063E ARABIC LETTER FARSI YEH WITH TWO DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063F ARABIC LETTER FARSI YEH WITH THREE DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+0640 ARABIC TATWEEL */
+  {0xF2AFu, 0xF2B0u, 0xF2B1u, 0xF2B2u}, /* U+0641 ARABIC LETTER FEH */
+  {0xF2B3u, 0xF2B4u, 0xF2B5u, 0xF2B6u}, /* U+0642 ARABIC LETTER QAF */
+  {0xF2B7u, 0xF2B8u, 0xF2B9u, 0xF2BAu}, /* U+0643 ARABIC LETTER KAF */
+  {0xF2BBu, 0xF2BCu, 0xF2BDu, 0xF2BEu}, /* U+0644 ARABIC LETTER LAM */
+  {0xF2BFu, 0xF2C0u, 0xF2C1u, 0xF2C2u}, /* U+0645 ARABIC LETTER MEEM */
+  {0xF2C3u, 0xF2C4u, 0xF2C5u, 0xF2C6u}, /* U+0646 ARABIC LETTER NOON */
+  {0xF2C7u, 0xF2C8u, 0xF2C9u, 0xF2CAu}, /* U+0647 ARABIC LETTER HEH */
+  {0x0000u, 0x0000u, 0xF2CCu, 0xF2CBu}, /* U+0648 ARABIC LETTER WAW */
+  {0x0000u, 0x0000u, 0xF2D3u, 0xF2D4u}, /* U+0649 ARABIC LETTER ALEF MAKSURA */
+  {0xF2CDu, 0xF2CEu, 0xF2CFu, 0xF2D0u}, /* U+064A ARABIC LETTER YEH */
+};
+
+#define SHAPING_TABLE_PUA2_FIRST	0x0622u
+#define SHAPING_TABLE_PUA2_LAST		0x064Au
+#define SHAPING_TABLE_PUA2_LEN		(SHAPING_TABLE_PUA2_LAST - SHAPING_TABLE_PUA2_FIRST + 1)
+
+static const struct ligature_set_t ligature_table_pua2[] =
+{
+  { 0xF2BBu, {
+//  { 0xF2BCu, 0xF2C9u, 0xF201u }, /* LAM WITH LAM WITH HEH ISOLATED FORM */
+//  { 0xF2C0u, 0xF25Au, 0xF210u }, /* LAM WITH MEEM WITH HAH INITIAL FORM */
+//  { 0xF2C0u, 0xF256u, 0xF211u }, /* LAM WITH MEEM WITH JEEM INITIAL FORM */
```

Comment:
I still need to figure out if the current code can easily support 3-letter ligatures.